### PR TITLE
[pytx][cli] Allow printing out signals, csv files

### PR DESCRIPTION
**Pull request by Dcallies**
Summary
---------

This PR adds the ability to print out data to screen from the dataset command. This is designed to close a functionality gap with the old CLI that stored data in simple csv files that could be passed around.

Test Plan
---------

Added test

(Note - below `tx` is used as an alias for `threatexchange`

### Halp
```
$ threatexchange dataset --help
usage: threatexchange dataset [-h] [--rebuild-indices | --clear-indices | --signal-summary | --print-signals] [--only-signals SIGNAL_TYPE [SIGNAL_TYPE ...] | --only-content CONTENT_TYPE
                              [CONTENT_TYPE ...]] [--only-collabs NAME [NAME ...]] [--only-tags STR] [--signals-only] [--csv] [--print-zeroes]

    Introspect fetched data.

    Can print out contents in simple formats
    (ideal for sending to another system), or regenerate
    index files (ideal if distributing indices for some reason)

    Example commands:

    ```
    # Show total size of known signals
    $ threatexchange dataset

    # Show the size of one collaboration
    $ threatexchange dataset -c 'Sample Data'

    # Generate a simple CSV file for one collaboration for ease of distribution
    $ threatexchange dataset -c 'Sample Data' -P --csv > collab.csv
    ```
    

optional arguments:
  -h, --help            show this help message and exit
  --rebuild-indices, -r
                        rebuild indices from fetched data
  --clear-indices, -X   clear all indices
  --signal-summary      print summary in terms of signals (default action)
  --print-signals, -P   print signals to screen
  --only-signals SIGNAL_TYPE [SIGNAL_TYPE ...], -s SIGNAL_TYPE [SIGNAL_TYPE ...]
                        only consider signals of this type
  --only-content CONTENT_TYPE [CONTENT_TYPE ...], -C CONTENT_TYPE [CONTENT_TYPE ...]
                        only process signals for these content types
  --only-collabs NAME [NAME ...], -c NAME [NAME ...]
                        [-S|-P] only count items with this tag
  --only-tags STR, -t STR
                        [-S|-P] only count items with these tags
  --signals-only, -S    [-P] only print signals
  --csv                 [-P] store in csv format
  --print-zeroes, -z    when printing counts, print 0 values
```

### Some examples with sample data

```
$ tx dataset
pdq: 138
raw_text: 3
video_md5: 2
trend_query: 1
url: 1
url_md5: 1

$ tx dataset -P | tail -n5
'Sample Signals' pdq d33c9979cbab36ba44ed72aa1358d6a7ac928d4168963512fbb511643336c4cb WORTH_INVESTIGATING
'Sample Signals' pdq c4b54f9e32c67134c704ce16dc6d31873634b339c263c44cc9c7958f1a396fb5 WORTH_INVESTIGATING
'Sample Signals' pdq b0f815db7693c05a4b765a66584dd92cdb952c31b46335e124ce049c7eb82ddb WORTH_INVESTIGATING
'Sample Signals' url_md5 95c3a3934d38b06959d4f6d5bfe48a79 WORTH_INVESTIGATING
'Sample Signals' url https://developers.facebook.com/docs/threat-exchange/reference/apis/ WORTH_INVESTIGATING

$ tx dataset -PSsurl
https://developers.facebook.com/docs/threat-exchange/reference/apis/

$ tx dataset -P -S -s pdq | head -n 5
acecf3355e3125c8e24e2f30e0d4ec4f8482b878b3c34cdbdf063278db275992
8fb70f36e1c4181e82fde7d0f80138430e1e31f07b628e31ccbb687e87e1f307
36b4665bca0c91f6aecb8948e3381e57e509ae7210e3cd1bd768288e56a95af9
e875634b9df48df5bd7f1695c796287e8a0ec0603c0c478170fc9d0f81ea60f4
42869d32fff9b14c100759e17b7c204628f97efca264c007f5e9bdfc004f2a73

$ tx dataset -P --csv | tail -n3
Sample Signals,url_md5,95c3a3934d38b06959d4f6d5bfe48a79,WORTH_INVESTIGATING,
Sample Signals,url,https://developers.facebook.com/docs/threat-exchange/reference/apis/,WORTH_INVESTIGATING,
Sample Signals,video_md5,cab08b36195edb1a1231d2d09fa450e0,WORTH_INVESTIGATING,

$ tx dataset -P --csv  > signals.csv
```

### CODE REVIEW

File: python-threatexchange/threatexchange/cli/dataset_cmd.py [hunk 1]
```diff
@@ -255,3 +279,38 @@ def execute_generate_indices(self, settings: CLISettings) -> None:
...
+
+    def _print_csv(
+        self,
+        csvwriter: csv.DictWriter,
+        collab_name: str,
+        signal_type: SignalType,
+        signal_str: str,
+        metadata: FetchedSignalMetadata,
+    ) -> None:
+        agg = metadata.get_as_aggregate_opinion()
```
Comment (BarrettOlson) [inline python-threatexchange/threatexchange/cli/dataset_cmd.py:307]: blocking: I think this means using `signals_only` and `csv` will result in error.

File: python-threatexchange/threatexchange/cli/dataset_cmd.py [hunk 2]
```diff
@@ -96,58 +110,67 @@ def init_argparse(cls, settings: CLISettings, ap: ArgumentParser) -> None:
...
             action="store_true",
             help="[-P] only print signals",
         )
         ap.add_argument(
-            "--limit",
-            "-l",
+            "--csv",
             action="store_true",
-            help="[-P] only print this many records",
+            help="[-P] store in csv format",
```
Comment (BarrettOlson) [inline python-threatexchange/threatexchange/cli/dataset_cmd.py:120]: store -> output
The word store really confused me and I started looking for a filename arg/default

File: python-threatexchange/threatexchange/cli/tests/dataset_cmd_test.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+from threatexchange.cli.tests.e2e_test_helper import ThreatExchangeCLIE2eTest
+
+
+class DatasetCommandTest(ThreatExchangeCLIE2eTest):
```
Comment (BarrettOlson) [inline python-threatexchange/threatexchange/cli/tests/dataset_cmd_test.py:4]: Thanks for adding this test!
Is there are reason you skipped using the cvs flag in one of them?

**Review by schatten — APPROVED** [review]

File: python-threatexchange/threatexchange/cli/dataset_cmd.py [hunk 4]
```diff
@@ -96,58 +110,67 @@ def init_argparse(cls, settings: CLISettings, ap: ArgumentParser) -> None:
         )
```
Comment (BarrettOlson) [inline python-threatexchange/threatexchange/cli/dataset_cmd.py:110]: question about existing command: only-tags takes a sequence? can you add some help text for how to delineate each?

**Review by BarrettOlson — APPROVED** [review]
LGTM, had a few nit and one blocker that I think it a small bug

File: python-threatexchange/threatexchange/cli/dataset_cmd.py (hunk 4 above)
Comment (Dcallies) [reply python-threatexchange/threatexchange/cli/dataset_cmd.py:110]: This flag may not actually work, will remove if so

File: python-threatexchange/threatexchange/cli/dataset_cmd.py (hunk 1 above)
Comment (Dcallies) [reply python-threatexchange/threatexchange/cli/dataset_cmd.py:307]: Ah, that's why I had that argument listed as optional originally.

File: python-threatexchange/threatexchange/cli/dataset_cmd.py (hunk 4 above)
Comment (Dcallies) [reply python-threatexchange/threatexchange/cli/dataset_cmd.py:110]: Looks like it is in use, but has never worked.

setting nargs=+ generates the default help text

**Pull request merged**